scalar_udf.go

```diff
@@ -0,0 +1,241 @@
+		C.duckdb_destroy_logical_type(&t)
+		return nil
+	}
+
+	// Set normal input parameters.
+	if config.InputTypeInfos() == nil {
+		return errScalarUDFNilInputTypes
+	}
+	if len(config.InputTypeInfos()) == 0 {
+		return errScalarUDFEmptyInputTypes
+	}
+
+	for i, info := range config.InputTypeInfos() {
+		if info == nil {
+			return addIndexToError(errScalarUDFInputTypeIsNil, i)
+		}
+		t := info.logicalType()
+		C.duckdb_scalar_function_add_parameter(f, t)
+		C.duckdb_destroy_logical_type(&t)
+	}
+	return nil
+}
+
+func registerResultParams(config ScalarFuncConfig, f C.duckdb_scalar_function) error {
+	if config.ResultTypeInfo() == nil {
+		return errScalarUDFResultTypeIsNil
+	}
+	if config.ResultTypeInfo().InternalType() == TYPE_ANY {
+		return errScalarUDFResultTypeIsANY
+	}
+	t := config.ResultTypeInfo().logicalType()
+	C.duckdb_scalar_function_set_return_type(f, t)
+	C.duckdb_destroy_logical_type(&t)
+	return nil
+}
+
+func createScalarFunc(name string, f ScalarFunc) (C.duckdb_scalar_function, error) {
+	if name == "" {
+		return nil, errScalarUDFNoName
+	}
+	if f == nil {
+		return nil, errScalarUDFIsNil
+	}
+	function := C.duckdb_create_scalar_function()
+
+	// Set the name.
+	cName := C.CString(name)
+	C.duckdb_scalar_function_set_name(function, cName)
+	C.duckdb_free(unsafe.Pointer(cName))
+
+	// Configure the scalar function.
+	config := f.Config()
+	extraInfo := f.ExtraInfo()
+	if err := registerInputParams(config, extraInfo, function); err != nil {
+		return nil, err
+	}
+	if err := registerResultParams(config, function); err != nil {
+		return nil, err
+	}
+	if extraInfo != nil && extraInfo.SpecialNullHandling() {
+		C.duckdb_scalar_function_set_special_handling(function)
+	}
+	if extraInfo != nil && extraInfo.Volatile() {
+		C.duckdb_scalar_function_set_volatile(function)
+	}
+
+	// Set the function callback.
+	C.duckdb_scalar_function_set_function(function, C.scalar_udf_callback_t(C.scalar_udf_callback))
+
+	// Set data available during execution.
+	h := cgo.NewHandle(f)
+	C.duckdb_scalar_function_set_extra_info(
+		function,
+		unsafe.Pointer(&h),
+		C.duckdb_delete_callback_t(C.scalar_udf_delete_callback))
+
+	return function, nil
+}
+
+// RegisterScalarUDF registers a scalar user-defined function.
+// The function takes ownership of f, so you must pass it as a pointer.
+func RegisterScalarUDF(c *sql.Conn, name string, f ScalarFunc) error {
+	scalarFunc, err := createScalarFunc(name, f)
+	if err != nil {
+		return getError(errAPI, err)
+	}
+
+	// Register the function on the underlying driver connection exposed by c.Raw.
+	err = c.Raw(func(driverConn any) error {
+		con := driverConn.(*conn)
+		state := C.duckdb_register_scalar_function(con.duckdbCon, scalarFunc)
+		C.duckdb_destroy_scalar_function(&scalarFunc)
+		if state == C.DuckDBError {
+			return getError(errAPI, errScalarUDFCreate)
+		}
+		return nil
+	})
+	return err
+}
+
+func RegisterScalarUDFSet(c *sql.Conn, name string, functions ...ScalarFunc) error {
```

Comment:
Although I like the API ATM, I think this will cause issues when adding chunked and threaded functions (I assume those also exist for scalar functions). How would you combine these into a single UDF set? This API could be changed to allow `... any` as paramter but that would also be bad.
I think it would be more future-proof to make the user first create the set, then add them one by one, and then register to duckdb. This way a different function can be used for adding chunk scalar functions and row scalar functions (using generics or not, doesn't matter).

Reply:
That's a good catch; I'll add the changes.
Also, I'll add a chunked row execution example + interface similar to your PR.

Reply:
For scalar functions, we always use chunks in the execution callback. We mirror duckdb's internal chunked execution engine here, so there are no different scalar function types, as is the case with table functions.

I've changed the API to allow adding more callback function types without creating a breaking change.
If, at any point, we want to expose `type chunkFn func(chunk DataChunk) (any, error)`, we can do so, but it might not become necessary. What do you think?
```Go
type rowFn func(args []driver.Value) (any, error)

type ScalarFuncExecutor struct {
	RowExecutor rowFn
}

type ScalarFunc interface {
	Config() ScalarFuncConfig
	Executor() ScalarFuncExecutor
}
```

Reply:
Sure this seems fine, I think I prefer how it is implemented for table functions but any way to add this functionality without a breaking change is good.